sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala

```diff
@@ -2148,6 +2149,17 @@ object RewriteIntersectAll extends Rule[LogicalPlan] {
   }
 }
 
+/**
+ * Deduplicate the right side of left-semi join and left-anti join.
+ */
+object DeduplicateLeftSemiLeftAntiRightSide extends Rule[LogicalPlan] {
+  def apply(plan: LogicalPlan): LogicalPlan = plan.transformWithPruning(
+    _.containsPattern(LEFT_SEMI_OR_ANTI_JOIN), ruleId) {
+    case join @ Join(_, right, LeftSemiOrAnti(_), _, _) if !right.isInstanceOf[Aggregate] =>
+      join.copy(right = Aggregate(right.output, right.output, right))
```

Comment:
Deduplicate is not always has benefit. This is my initial PR: https://github.com/apache/spark/pull/33465/commits/b5599010ba969ba4cc3a2ce85549fe226b75ae65

Reply:
yes, there are some cases that had not been taken into account in this PR. I think https://github.com/apache/spark/commit/b5599010ba969ba4cc3a2ce85549fe226b75ae65 is much better. I will close this one.